### PR TITLE
Multicall support and account client refactor

### PR DESCRIPTION
**Pull request by war-in**
### **Please check if the PR fulfills these requirements**
- [x] The formatter, linter, and tests all run without an error
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

## **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

It adds support for multicalls.

There are some new methods in AccountClient:
- prepare_invoke_function
- sign_transaction
- send_transaction
- execute

## **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

Yes, add_transaction was removed from client interface

### CODE REVIEW

**Review by cptartur — CHANGES_REQUESTED** [review]
I left some comments, AccountClient methods could use some simplification

Also guide and quickstart should be updated to show usages of the new `.execute` and explanation why it's there.

File: starknet_py/net/account/account_client.py [hunk 1]
```diff
@@ -389,24 +393,25 @@ def add_signature_to_transaction(
...
-    ):
-        calldata.append(
-            {
-                "to": call.to_addr,
-                "selector": call.selector,
-                "data_offset": current_data_len,
-                "data_len": len(call.calldata),
-            }
-        )
+    def parse_call(call: Call, current_data_len: int, entire_calldata: List):
```
Comment (cptartur) [inline starknet_py/net/account/account_client.py:396]: Maybe add a return type annotation?

File: starknet_py/contract.py [hunk 2]
```diff
@@ -194,45 +196,28 @@ async def call(
...
         """
-        if auto_estimate and max_fee is not None:
+        if not isinstance(self._client, AccountClient):
             raise ValueError(
-                "Max_fee and auto_estimate are exclusive and cannot be provided at the same time."
-            )
-        if auto_estimate and self.max_fee is not None:
-            raise ValueError(
-                "Auto_estimate cannot be used if max_fee was provided when preparing a function call."
+                "Use the AccountClient while creating Contract instance to invoke the transaction"
```
Comment (Solpatium) [inline starknet_py/contract.py:211]: I think something like "Contract uses an account that can't invoke transactions. You need to use AccountClient for that." explains it better.

File: starknet_py/contract.py [hunk 3]
```diff
@@ -241,25 +226,30 @@ async def invoke(
...
 
-        :param signature: Signature to send
         :return: Estimated amount of Wei executing specified transaction will cost
         :raises ValueError: when max_fee of PreparedFunctionCall is not None or 0.
         """
+        if not isinstance(self._client, AccountClient):
+            raise ValueError(
+                "Use the AccountClient while creating Contract instance to estimate fee of the transaction"
+            )
+
```
Comment (Solpatium) [inline starknet_py/contract.py:245]: Repetition. Please extract to `assert_can_invoke`.

File: starknet_py/net/account/account_client.py [hunk 4]
```diff
@@ -171,52 +184,128 @@ async def get_balance(
 
         return (high << 128) + low
 
-    async def _prepare_execute_transaction(self, tx: InvokeFunction) -> Transaction:
+    async def prepare_invoke_function(
```
Comment (Solpatium) [inline starknet_py/net/account/account_client.py:187]: `InvokeFunction` is a terrible name considering what is really happening. Can we brainstorm a better name and use it as alias for `InvokeFunction`?

**Review by Solpatium — CHANGES_REQUESTED** [review]

File: starknet_py/net/account/account_client.py (hunk 4 above)
Comment (war-in) [reply starknet_py/net/account/account_client.py:187]: We decided that we'll discuss it later because we couldn't find a solution

**Review by Solpatium — APPROVED** [review]

**Review by Arcticae — APPROVED** [review]

**Review by cptartur — APPROVED** [review]

File: docs/migration_guide.rst [hunk 5]
```diff
@@ -61,4 +62,12 @@ Interface of :ref:`Contract` remains unchanged and it is still the recommended w
 AccountClient
 -------------
 
-:ref:`AccountClient` now implements ``Client`` ABC: parameters of some of its' methods has changed.
\ No newline at end of file
+:ref:`AccountClient` now implements ``Client`` ABC: parameters of some of its' methods has changed.
```
Comment (Solpatium) [inline docs/migration_guide.rst:65]: now implements ``Client`` **interface**: parameters of some of its methods **were** changed.

**Pull request merged**